### ISSUE TITLE
Running 'pnpm install ... --use-store-server ...' in parallel is unsafe

### ISSUE THREAD

**Issue by jbrekelmans** [pnpm/pnpm#1148]
# Details
Any platform.
pnpm version v1.41.3.
# Detailed description of failure
Consider the following scenario: 
No pnpm store server is running and two 'pnpm install --use-store-server' processes, say A and B, are run in parallel.
Both A and B observe there is no server.json file and start a server. 
A's server writes server.json. 
A reads the server.json.
B's server writes server.json.
B reads the server.json.
A performs installation via A's server and simultaneously B performs installation via B's server.
The same store is modified without any locking.
This results in installation errors like the following:
```
ERROR  ENOTEMPTY: directory not empty, rename '.pnpm-store/2/registry.npmjs.org/@webassemblyjs/helper-wasm-bytecode/1.3.1/_tmp_28359_c59f0c6972e27ecda4311ecafc762f90' -> '.pnpm-store/2/registry.npmjs.org/@webassemblyjs/helper-wasm-bytecode/1.3.1/node_modules/@webassemblyjs/helper-wasm-bytecode'
    at 1/node_modules/
```
# Workarounds
1. Ensure no two 'pnpm install ... --use-store-server ...' commands run in parallel, unless you know the server is running.
2. Ensure --lock is used for all installations that may run in parallel with another.  
3. Ensure 'pnpm install' does not need to be run in parallel (e.g. 'pnpm recursive ...')

# Why?
**Why should 'pnpm install ... --use-store-server ...' guard against parallel starting of servers?**
'pnpm install ...' should 'just work', even when running installs in parallel. The use case I have is as follows:
I have a CI job that builds a number of packages in a monolithic repository. I would rather not use one of 
the above workarounds because it introduces complexity and noise in the CI job's scripts (or slows down the job).
In other words: how 'pnpm install ...' works should be a hidden implementation detail.
